core/trino-main/src/main/java/io/trino/operator/TopNOperator.java

```diff
@@ -158,18 +171,33 @@ private TopNOperator(
         return pages;
     }
 
+    private boolean isBuilderFull()
+    {
+        return maxPartialMemory <= topNProcessor.getEstimatedSizeInBytes();
+    }
+
+    private void addPage(Page page)
+    {
+        checkState(!isPartialFlushing, "TopN buffer is already full");
+        topNProcessor.addInput(page);
+        if (isBuilderFull()) {
+            isPartialFlushing = true;
+        }
+    }
+
     private class TopNPages
             implements WorkProcessor.Transformation<Page, Page>
     {
         @Override
         public TransformationState<Page> process(Page inputPage)
         {
-            if (inputPage != null) {
-                topNProcessor.addInput(inputPage);
-                return TransformationState.needsMoreData();
+            if (!isPartialFlushing && inputPage != null) {
+                addPage(inputPage);
```

Comment:
nit: you can inline `addPage`. It's only used once

Reply:
I would prefer to have it as a separate method.